karrot/activities/renderers.py

```diff
@@ -0,0 +1,44 @@
+from rest_framework import renderers
+from rest_framework.utils.serializer_helpers import ReturnDict
+
+
+class ICSCalendarRenderer(renderers.BaseRenderer):
+    """renders a list of iCal VEVENT as an iCal VCALENDAR."""
+    media_type = 'text/calendar'
+    format = 'ics'
+    charset = 'utf-8'
+
+    def render(self, data, media_type=None, renderer_context=None):
+        renderer_context = renderer_context or {}
+        response = renderer_context['response']
+        if response.exception:
+            return response.status_text.title()
+
+        # render lists of results as one calendar
+        if isinstance(data, ReturnDict):
+            results = [data]
+        else:
+            results = data
+
+        lines = [
+            'BEGIN:VCALENDAR',
+            'VERSION:2.0',
+            'PRODID:-//Karrot//EN',
+        ]
+        lines += [self.render_vevent(vevent) for vevent in results]
+        lines += ['END:VCALENDAR']
+        # CRLF is required by the specs
+        return '\r\n'.join(lines)
+
+    def render_vevent(self, vevent):
+        """renders a single event"""
+        lines = [
+            'BEGIN:VEVENT',
+        ]
+        lines += [
+            '{}:{}'.format(key.upper().replace('_', '-'), value.replace('\n', '\\n')) for key, value in vevent.items()
+        ]
+        lines += [
+            'END:VEVENT',
+        ]
+        return '\r\n'.join(lines)
```

Comment:
What do you think about using a library to handle the ics formatting? https://icspy.readthedocs.io/en/stable/ looks reasonable? Is there a specific reason to have handrolled rendering?

Reply:
its a simple format and i wasnt sure how happy you are with new deps. but i was considering https://pypi.org/project/icalendar/ actually. also not sure if its natural to make it work with drf nice? what architecture would you use?

Reply:
In general I'm happier with libraries (maintained by somebody else) than hand-rolled stuff (have to somehow keep it my head), and the one you mentioned looks good too (more recently released as well).

Looks like it would be easy to use with your new activity serializer too... I think you could just replace the code inside your custom renderer with a call to the library. Does it make sense?